### ISSUE TITLE
PhantomJS Error - incompatible type of argument(s) in call to _uploadFile()

### ISSUE THREAD

**Issue by aaaskew** [teodesian/Selenium-Remote-Driver#341]
I am using:

Perl 5.14.2
Selenium Remote Driver 1.20
PhantomJS 2.1.1 linux x86_64 (binary from web site)

Call to $driver->upload_file( $fname ); fails:

```
REQ: POST, http://127.0.0.1:8910/wd/hub/session/e0f72710-b7ef-11e7-a963-9dee75ac455b/file, {"file":"UEsDBBQA...ABGBQAAAAA="}
RES: Error - incompatible type of argument(s) in call to _uploadFile(); candidates were
    _uploadFile(QString,QStringList)
```

Is PhantomJS no longer compatible with the upload_file method?

I see discussion on https://github.com/detro/ghostdriver/issues/282 suggesting:

call the POST /session/<session_id>/file endpoint with the following json
```
{
  "selector": "input[name=file]",
  "filepath": ["file1", "file2"],
}
```
or call the POST /session/<session_id>/phantom/execute endpoint with the following json
```
{
  "script": "self.uploadFile('input[name=file]', ['file1', 'file2']);",
  "args": [],
}
```

The following comment seems to explain the issue: https://github.com/ariya/phantomjs/issues/13302#issuecomment-158628631 .  Can the Selenium-Remote-Driver send one of these to PhantomJS maybe with a hidden method or data as a workaround?  I don't think that the issue I describe is the same as the linked multiple file issue although the modem web page I am trying to operate on has a weird set of forms. A summary follows:

```
<form action="../x509LoCfIm.cgi" enctype="multipart/form-data" id="frm1" method="post" name="frm1" target="XLOCFSender">
  <input class="file" name="srcfile" size="30" type="file" />
  <input name="attach" onclick="return waitResponse(0)" type="submit" value="Import" />
  <input class="btnw" name="btnCancel" onclick="btnCcl()" type="button" value="Cancel" />

<form action="../x509LoCfIm.cgi" enctype="multipart/form-data" id="frm2" method="post" name="frm2" target="XLOCFSender">
  <input class="file" name="p12file" size="30" type="file" />
  <input maxlength="23" name="dpwd" size="22" type="password" />
  <input name="attach" onclick="return waitResponse(1)" type="submit" value="Import" />
  <input class="btnw" name="btnCancel" onclick="btnCcl()" type="button" value="Cancel" />

<form action="../x509LoCfIm.cgi" enctype="multipart/form-data" id="frm3" method="post" name="frm3" target="XLOCFSender">
  <input class="file" name="crtfile" size="30" type="file" />
  <input class="file" name="keyfile" size="30" type="file" />
  <input maxlength="23" name="dpwd" size="22" type="password" />
  <input name="attach" onclick="return waitResponse(2)" type="submit" value="Import" />
  <input class="btnw" name="btnCancel" onclick="btnCcl()" type="button" value="Cancel" />

```

An extract of the code I am using:

```
my $driver = Selenium::PhantomJS->new( binary            => '/usr/local/bin/phantomjs',
                                       base_url          => 'https://modem:443/');
$driver->get('/');
...
$driver->switch_to_frame('XLOCFIMPORT');
my $webelement = $driver->find_element_by_xpath('//*[./form[@name="frm3"]]');

$driver->debug_on;
my $remote_crtfname = $driver->upload_file( $crtfname );   #  <-- Dies here
my $crtelement = $driver->find_child_element($webelement, './/*[@name="crtfile"]');
$crtelement->send_keys( $remote_crtfname );

```

**Comment by teodesian** [comment]
Given the upload file in mainline selenium is an "unofficial" endpoint, none of this shocks me, and yes a special handler will have to be made given that phantom has no maintenance anymore, and no intention of fixing:

https://github.com/facebook/php-webdriver/issues/458

I really need to take time to work on headless chromedriver...

I'll look at a patch soonish.

**Comment by Qarj** [comment]
At the Selenium conference in Berlin they said PhantomJS has been abandoned and they recommended against using it.

My experience with PhantomJS is that is worked ok for some pages, and for other pages it crashes so I gave up long ago.

With headless ChromeDriver it seemed to work really well at first - but sometimes we get totally blank pages so we've stopped using that too. Maybe it is only a Windows problem.
